src/components/text-editor/prosemirror-adapter/plugins/trigger/factory.ts

```diff
@@ -0,0 +1,227 @@
+    );
+};
+
+const getTriggerEventDetail = (
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+): TriggerEventDetail => {
+    return {
+        trigger: trigger,
+        textEditor: inserterFactory(view),
+        value: value,
+    };
+};
+
+const sendTriggerEvent = (
+    type: 'triggerStart' | 'triggerStop' | 'triggerChange',
+    view: EditorView,
+    trigger: TriggerCharacter,
+    value: string,
+) => {
+    const event = new CustomEvent<TriggerEventDetail>(type, {
+        detail: getTriggerEventDetail(view, trigger, value),
+        bubbles: true,
+        composed: true,
+    });
+    view.dom.dispatchEvent(event);
+};
+
+const processTransactions = (
+    text: string,
+    transactions: Transaction[],
+    oldState: EditorState,
+): string => {
+    let textAdded = '';
+    let textRemoved = '';
+
+    transactions.forEach((transaction) => {
+        if (transaction.docChanged) {
+            transaction.steps.forEach((step) => {
+                if (
+                    step instanceof ReplaceStep ||
+                    step instanceof ReplaceAroundStep
+                ) {
+                    const slice = step.slice;
+                    const fromPos = step.from;
+                    const toPos = step.to;
+
+                    if (slice?.size) {
+                        // Text added
+                        textAdded += slice.content.textBetween(0, slice.size);
+                    } else if (fromPos !== toPos) {
+                        // Text removed
+                        const removedText = oldState.doc.textBetween(
+                            fromPos,
+                            toPos,
+                        );
+                        textRemoved += removedText;
+                    }
+                }
+            });
+        }
+    });
+
+    if (textAdded) {
+        text += textAdded;
+    } else if (textRemoved) {
+        text = text.slice(0, -textRemoved.length);
+    }
+
+    return text;
+};
+
+export const createTriggerPlugin = (triggerCharacters: TriggerCharacter[]) => {
+    let activeTrigger: TriggerCharacter | null = null;
+    let triggerText = '';
+    let pluginView: EditorView | null = null;
+    let triggerPosition: number | null = null;
+
+    const stopTrigger = () => {
+        triggerText = '';
+        sendTriggerEvent('triggerStop', pluginView, activeTrigger, triggerText);
+        triggerPosition = null;
+        activeTrigger = null;
+    };
+
+    const handleKeyDown = (view: EditorView, event: any) => {
+        const { state } = view;
+
+        if (event.key === 'Escape') {
+            stopTrigger();
+
+            return true;
+        }
+
+        if (isTrigger(event.key, triggerCharacters) && shouldTrigger(state)) {
+            activeTrigger = event.key;
+            triggerText = '';
+            triggerPosition = state.selection.$from.pos - triggerText.length;
+            sendTriggerEvent('triggerStart', view, activeTrigger, triggerText);
```

Comment:
I came across this when working on https://github.com/Lundalogik/lime-elements/pull/3307. What's `triggerText` for, and is it supposed to ever be anything other than an empty string? If not, we can make the following change:

```diff
             triggerText = '';
-            triggerPosition = state.selection.$from.pos - triggerText.length;
+            triggerPosition = state.selection.$from.pos;
             sendTriggerEvent('triggerStart', view, activeTrigger, triggerText);
```

On the other hand, since it's a variable everywhere, it kinda looks like the intention is that it's not always going to remain an empty string? 🙂

Reply:
We used it on line 193 in this file. :) 

````ts
        if (updatedText !== triggerText) {
            triggerText = updatedText;
            sendTriggerEvent(
                'triggerChange',
                pluginView,
                activeTrigger,
                triggerText.slice(1),
            );
        }
````